litgpt/model.py

```diff
@@ -253,8 +258,22 @@ def forward(
 
     def scaled_dot_product_attention(
         self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, mask: Optional[torch.Tensor] = None
-    ) -> torch.Tensor:
+    ):
         scale = 1.0 / math.sqrt(self.config.head_size)
+        if (
+            FlashAttention3Available
+            and mask is None
+            and q.device.type == "cuda"
+            and q.dtype in (torch.float16, torch.bfloat16)
+            and has_h100_or_h800()
+        ):
+            from flash_attn import flash_attn_func
+
+            # flash-attn requires (B, T, nh, hs)
+            q = q.transpose(1, 2)
+            k = k.transpose(1, 2)
+            v = v.transpose(1, 2)
+            return flash_attn_func(q, k, v, dropout_p=0.0, softmax_scale=scale, causal=True)
```

Comment:
Have you vetted this code or just copied it from one of the old commits?
I remember that previously `LitGPT` used both `torch SDPA` and `flash_attn`, so maybe you took this code from there.

There are a couple of questions that I would like to have answers to before the PR is merged:
1. `mask`. Before a mask disabled flash attention internally, but it was back then. Is it still true?
2. `dtype`. I would assume, since FlashAttentionv3 has just been released, or even still in development, maybe it only support `bfloat16`?
3. do we still need to `transpose` q, k and v matrices

Reply:
Not vetted yet, I need to test this more. The challenge is that it only works on H100. The limited time I had testing it, it was 5x slower than not using it.

Before merging, we should definitely make a benchmark table with some concrete numbers.

> perhaps it only supports bfloat16

It's the other way around and only supports float16; in the README they said bfloat16 was coming soon. 

 > do we still need to transpose q, k and v matrices

This still needs investigation